iroh/examples/transfer.rs

```diff
@@ -0,0 +1,313 @@
+            let (mut send, mut recv) = conn.accept_bi().await?;
+            tracing::debug!("accepted bi stream, waiting for data...");
+            let message = recv.read_to_end(100).await?;
+            let message = String::from_utf8(message)?;
+            println!("received: {message}");
+
+            send_data_on_stream(&mut send, size).await?;
+
+            // We sent the last message, so wait for the client to close the connection once
+            // it received this message.
+            let res = tokio::time::timeout(Duration::from_secs(3), async move {
+                let closed = conn.closed().await;
+                if !matches!(closed, quinn::ConnectionError::ApplicationClosed(_)) {
+                    println!("node {node_id} disconnected with an error: {closed:#}");
+                }
+            })
+            .await;
+            if res.is_err() {
+                println!("node {node_id} did not disconnect within 3 seconds");
+            }
+            Ok::<_, anyhow::Error>(())
+        });
+    }
+
+    // stop with SIGINT (ctrl-c)
+    Ok(())
+}
+
+async fn fetch(ticket: &str, relay_url: Option<String>) -> anyhow::Result<()> {
+    let ticket: NodeTicket = ticket.parse()?;
+    let secret_key = SecretKey::generate();
+    let relay_mode = match relay_url {
+        Some(relay_url) => {
+            let relay_url = RelayUrl::from_str(&relay_url)?;
+            let relay_map = RelayMap::from_url(relay_url);
+            RelayMode::Custom(relay_map)
+        }
+        None => RelayMode::Default,
+    };
+    let endpoint = Endpoint::builder()
+        .secret_key(secret_key)
+        .alpns(vec![TRANSFER_ALPN.to_vec()])
+        .relay_mode(relay_mode)
+        .bind()
+        .await?;
+
+    let start = Instant::now();
+
+    let me = endpoint.node_id();
+    println!("node id: {me}");
+    println!("node listening addresses:");
+    for local_endpoint in endpoint
+        .direct_addresses()
+        .next()
+        .await
+        .context("no endpoints")?
+    {
+        println!("\t{}", local_endpoint.addr)
+    }
+
+    let relay_url = endpoint
+        .home_relay()
+        .expect("should be connected to a relay server, try calling `endpoint.local_endpoints()` or `endpoint.connect()` first, to ensure the endpoint has actually attempted a connection before checking for the connected relay server");
+    println!("node relay server url: {relay_url}\n");
+
+    // Attempt to connect, over the given ALPN.
+    // Returns a Quinn connection.
+    let conn = endpoint
+        .connect(ticket.node_addr().clone(), TRANSFER_ALPN)
+        .await?;
+    info!("connected");
+
+    // Use the Quinn API to send and recv content.
+    let (mut send, mut recv) = conn.open_bi().await?;
+
+    let message = format!("{me} is saying 'hello!'");
+    send.write_all(message.as_bytes()).await?;
+
+    // Call `finish` to close the send side of the connection gracefully.
```

Comment:
```suggestion
    // Call `finish` to signal no more data will be sent on this stream.
```
This doesn't really have anything to do with gracefully closing anything.  It only signals that no more data will be sent on this stream.  (Which is what `read_to_end` on the other side relies on.)